### ISSUE TITLE
Difference in output between jitted and non-jitted call

### ISSUE THREAD

**Issue by rodrigodzf** [jax-ml/jax#20371]
### Description

Originally I posted this to the Flax repo(https://github.com/google/flax/issues/3777), but I think this is jax issue now.

I have found that the results of the forward pass, differ considerably if the **apply** function is jitted or not.

### System information
- Ubuntu 23.10 x86_64
- Flax, jax, jaxlib versions (obtain with `pip show flax jax jaxlib`: flax 0.8.2, jax 0.4.25, jaxlib 0.4.25+cuda12.cudnn89
- Python version: v3.10.13
- GPU/TPU model and memory:  NVIDIA GeForce RTX 3090
- CUDA version (if applicable): 12.4

### Problem you have encountered:

When using the forward pass for a simple MLP the results are different with the jitted version.

### What you expected to happen:

That the results are the same.

### Steps to reproduce:
```python
import flax.linen as nn
import jax.numpy as jnp
import jax
from typing import Sequence

class MLP(nn.Module):
    hidden_channels: Sequence[int]
    activation: nn.Module = nn.relu

    @nn.compact
    def __call__(self, x: jnp.ndarray):
        for i, channels in enumerate(self.hidden_channels):
            x = nn.Dense(features=channels)(x)
            if i != len(self.hidden_channels) - 1:
                x = self.activation(x)
        return x
    
d_hidden = 64
d_input = 64
d_batch = 3
proj = MLP(
    hidden_channels=[d_hidden] * 2,
    activation=nn.selu,
)

proj_vars = proj.init(jax.random.PRNGKey(546543), jnp.ones((d_batch, d_input)))
x = jnp.ones((d_batch, d_input)) 

out = proj.apply(proj_vars, x)
jitted_out = jax.jit(proj.apply)(proj_vars, x)
diff = jnp.abs(out - jitted_out).mean()
print(f"diff={diff}")
```
this gives an error of 0.0005452320910990238 in my computer. 

I found out that from `jaxlib==0.4.21` to `jaxlib==0.4.25` I get this error but with `jaxlib==0.4.20` there is no error.

Interestingly if I use  '--xla_gpu_triton_gemm_any=True ' there is no error as well.
 

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.25
jaxlib: 0.4.25
numpy:  1.26.4
python: 3.10.13 | packaged by conda-forge | (main, Dec 23 2023, 15:36:39) [GCC 12.3.0]
jax.devices (1 total, 1 local): [cuda(id=0)]
process_count: 1
platform: uname_result(system='Linux', node='home', release='6.5.0-26-generic', version='#26-Ubuntu SMP PREEMPT_DYNAMIC Tue Mar  5 21:19:28 UTC 2024', machine='x86_64')
```

**Comment by mattjj** [comment]
Thanks for the question!

I think this is working as intended: especially on GPU, floating point numerics may not be exactly preserved under compilation. Indeed, floating point numerics may not even be preserved run-to-run on GPU, unless you set some flags that significantly degrade performance! See https://github.com/google/jax/discussions/10674#discussioncomment-7214817 about nondeterminism.

I'm not sure exactly why that flag would affect things, other than it's just changing the generated code.

> differ considerably

0.0005452320910990238 does not seem surprising to me for f32, but if it's too much for your use case, maybe try a higher floating point precision (since presumably the compiler-induced numerical noise scales down with increased precision)? Or it may be that one can identify XLA:GPU flags that are best for ensuring jit invariance, though I'm not sure if those will be stable over time...

WDYT?

**Comment by rodrigodzf** [comment]
Thanks for the prompt answer!

I think that makes sense, but what I don't understand is why I get a different behaviour with `jaxlib==0.4.20` vs `jaxlib==0.4.21+`  or why the  '--xla_gpu_triton_gemm_any=True ' seems to fix the problem as well. Also if I run this in a collab notebook with 0.4.23 then I get no numerical error

**Comment by jakevdp** [comment]
@mattjj is it possible that device-specific behavior for [`default_matmul_precision`](https://jax.readthedocs.io/en/latest/_autosummary/jax.default_matmul_precision.html) could account for the difference across GPU types?

**Comment by mattjj** [comment]
@jakevdp good point! Yeah I'm not sure, but different GPU types would also affect codegen in other ways.

**Comment by rajasekharporeddy** [comment]
Hi @rodrigodzf

The repro now produces a difference of 0.0 when run on RTX 2060 in WSL with JAX 0.4.37. Please find the attached screenshots for reference.

![image](https://github.com/user-attachments/assets/3d862ae5-97cf-416d-821f-bf72dd773524)

jax.print_environment_info():
![image](https://github.com/user-attachments/assets/9b9f68f4-09d1-4874-ba60-41401e651f33)

Could you please verify if the issue still persists in your system and let us know.

Thank you.

**Comment by rodrigodzf** [comment]
Hi, 

With versions >=0.4.37, the issue seems to be fixed. Thanks!
